futures-util/src/stream/shared.rs

```diff
@@ -0,0 +1,94 @@
+use crate::future::{Shared as SharedFuture, FutureExt};
+use futures_core::future::Future;
+use futures_core::stream::Stream;
+use futures_core::task::{self, Poll};
+use std::marker::Unpin;
+use std::mem::PinMut;
+use super::StreamExt;
+use super::into_future::StreamFuture;
+
+/// A stream that is cloneable and can be polled in multiple threads.
+/// Use `Stream::shared()` method to convert any stream into a `Shared` stream.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct Shared<St: Stream + Unpin>
+where
+    St::Item: Clone,
+{
+    inner: SharedFuture<Inner<St>>,
+}
+
+impl<St: Stream + Unpin> Unpin for Shared<St>
+where
+    St::Item: Clone,
+{}
+
+struct Inner<St: Stream + Unpin>(StreamFuture<St>) where St::Item: Clone;
+
+impl<St> Future for Inner<St>
+where
+    St: Stream + Unpin,
+    St::Item: Clone,
+{
+    type Output = Option<(St::Item, SharedFuture<Self>)>;
+
+    fn poll(mut self: PinMut<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        let (item, stream) = ready!(PinMut::new(&mut self.0).poll(cx));
+        Poll::Ready(item.map(|it| (it, Inner(stream.into_future()).shared())))
```

Comment:
Internally, calling `shared` allocates a new `Arc`, so this would result in a new allocation for every item in the stream (separately from the `Clone` that already has to happen for every item in the stream). Essentially, it builds a linked-list of the elements yielded by the stream. I'd prefer instead if the combinator kept a vector of all the elements yielded so far that have not yet been consumed by an interested waker, though I'm not exactly sure of the details. What do you think? Does this seem achievable? If not, I'd still be interested in adding support for this, but I'm concerned about the cost of all the allocations required and would want to include a note about them in the docs for the method.

Reply:
>  Essentially, it builds a linked-list of the elements yielded by the stream. 

Yep, that's exactly how this works. My first attempt didn't use the future's shared method, but it also created a shared-structure linked list of Arcs. I eventually put that approach on ice because it was pretty complicated and used a fair amount of unsafe code.

> I'd prefer instead if the combinator kept a vector of all the elements yielded so far that have not yet been consumed by an interested waker

If I understand correctly, that would have different behavior than this though. With this implementation, every copy of the stream will get the same items in the same order. If items are only kept until they are consumed by one waker, then each item will only every be seen by one copy of the stream.

I think there are two other possibilities:
1. Keep a Vec of all results that have been retrieved, and each copy of the stream keeps track of a position in the Vec, once one stream reaches the end, it calls `poll_next`. The problem with this is that it either needs to continue to keep items in the vec after they are no longer needed, or somehow keep of all copies of the stream and where they are so that when an item is no longer needed it can be removed from the Vec, and the indexes stored in all streams updated.
2. Using a slab instead of Arcs. That would probably require re-implementing some of Arcs logic, but with slab references instead of pointers.

Reply:
There probably should the possibility to optionally specify a capacity limit for the vector / `VecDeque` that stores the items

Reply:
@tmccombs 
> If I understand correctly, that would have different behavior than this though. With this implementation, every copy of the stream will get the same items in the same order. If items are only kept until they are consumed by one waker, then each item will only every be seen by one copy of the stream.

Yeah, my statement 
> kept a vector of all the elements yielded so far that have not yet been consumed by an interested waker

could have been clearer: I meant to say "a vector of all elements yielded so far *for which there exists a consumer that has not yet seen them*". So what you'd need to do would be to track the number of consumers that have not yet seen each value, and only pull them off of the queue for good once everyone has seen them.

Reply:
@tmccombs did you have any luck getting this going?

Reply:
I've made some progress on it. I've been fairly busy with other things, and getting this working is substantially more complicated than my first attempt. I definitely think it is feasible, though.

Reply:
I've pushed a branch up at https://github.com/tmccombs/futures-rs/tree/shared-stream-vec  I haven't quite got it working yet, but I'd like feedback on my approach.